### DISCUSSION TITLE
How to create perforation?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1438] [Q&A]
Hi Experts,

    Do we have any documentation about those attributes of creating perforation? I tried, but couldn't figure out how to make it work in my case.
````
My example,  
    <InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="Mesh"
      polylineNodeCoords="{ { 32441.418, 79301.3881, 685.94},
                            { 32441.418, 79301.3881, 657.75} }"
      polylineSegmentConn="{ { 0, 1 } }"
      radius="0.09525"
      numElementsPerSegment="1">
      <Perforation
        name="injector1_perf1"
        distanceFromHead="1"/>
    </InternalWell>
````
and GEOSX gives me error:
````
***** ERROR
***** LOCATION: /apps/GEOSX/src/coreComponents/mesh/WellElementRegion.cpp:67
***** Controlling expression (should be false): matchedPerforations != numPerforationsGlobal
***** Rank 0: Invalid mapping perforation-to-element in well wellRegion. This happens when GEOSX cannot match a perforation with a reservoir element. The most common reason for this error is that a perforation is on a section of  the well polyline located outside the domain.
````
My reservoir depth: 412 to 690(17 layers) and one perforation from 657.75 to 685.94(layer 12 to 16). I guess my understanding about the following attributes are wrong:
**polylineNodeCoords:, polylineSegmentConn, numElementsPerSegment, distanceFromHead**

what is the right format to create a perforation in my case? 

Thanks and Regards,

wdm

**Comment by wdmhouston** [comment]
Double checked my GRDECL file , updated my model and changed depth to below MSL,
````
    <InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="mesh"
      polylineNodeCoords="{ { 4983.49, 4983.49, -400.00},
                            { 4983.49, 4983.49, -685.94} }"
      polylineSegmentConn="{ { 0, 1 } }"
      radius="0.09525"
      numElementsPerSegment="2">
      <Perforation
        name="injector1_perf1"
        distanceFromHead="12.67"/>
    </InternalWell>
````
but GEOSX still reports error:
```
Writing into the GEOSX mesh data structure
***** ERROR
***** LOCATION: /apps/GEOSX/src/coreComponents/mesh/WellElementRegion.cpp:67
***** Controlling expression (should be false): matchedPerforations != numPerforationsGlobal
***** Rank 0: Invalid mapping perforation-to-element in well wellRegion. This happens when GEOSX cannot match a perforation with a reservoir element. The most common reason for this error is that a perforation is on a section of  the well polyline located outside the domain.

** StackTrace of 6 frames **
Frame 0: geosx::ElementRegionManager::generateWells(geosx::MeshManager&, geosx::MeshLevel&)
Frame 1: geosx::ProblemManager::generateMesh()
Frame 2: geosx::ProblemManager::problemSetup()
Frame 3: geosx::GeosxState::initializeDataRepository()
Frame 4: main
Frame 5: __libc_start_main
Frame 6: _start
=====
````

**Reply by francoishamon** [reply]
Ok, usually this happens when the perforation does not fall in the reservoir, but it is hard to tell without looking at your mesh. Can you try the following things, and report what works/ does not work:

Option 1:
```
<InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="mesh"
      polylineNodeCoords="{ { 4983.49, 4983.49, -400.00},
                            { 4983.49, 4983.49, -685.94} }"
      polylineSegmentConn="{ { 0, 1 } }"
      radius="0.09525"
      numElementsPerSegment="1">
      <Perforation
        name="injector1_perf1"
        distanceFromHead="12.67"/>
    </InternalWell>
```

Option 2:
```
<PAMELAMeshGenerator
      name="mesh"
      file="theNameOfTheFile.GRDECL"
      reverseZ="1"/>
```
with 
```
<InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="mesh"
      polylineNodeCoords="{ { 4983.49, 4983.49, -400.00},
                            { 4983.49, 4983.49, -685.94} }"
      polylineSegmentConn="{ { 0, 1 } }"
      radius="0.09525"
      numElementsPerSegment="1">
      <Perforation
        name="injector1_perf1"
        distanceFromHead="12.67"/>
    </InternalWell>
```
If nothing works, I suggest you try with the internal mesh generator, we make it work, and then you come back to your GRDECL file

**Reply by wdmhouston** [reply]
Francois,

      Thanks for the update. I fixed it using mesh generator.

       btw, in GRDECL, the depth value in ZCORN is positive below MSL(mean sea level), so the depth outputted to VTK file should be negative(check the vtk output of spe10 example).

**Comment by francoishamon** [comment]
@wdmhouston Regarding your questions: 

_How is well perforation calculated?_ 
The well perforation is calculated using the standard Peaceman model. Skin is currently assumed to be zero.

_Can geosx output well perforation information(cell, h, Kh, skin, transmissibility) to log?_
Please post a feature request

_What is the start and end measure depth of each perforation? Do I need to add perforation for each cell along the well path?_
The field `distanceFromHead` is used to identify a position along the well trajectory and compute its (x,y,z) coordinates using the linear distance from the well head (= top of the well). Then, GEOSX identifies the reservoir cell located at this position, and introduces a connection (= a perforation) between this reservoir cell and the well. So, if you want the well to be connected to each reservoir cell along the well path, yes, you need to add a perforation for each cell.

_Transmissibility value is not accepted in current release(bug)_
Have you tried to use the `transmissibility` field in the `Perforation` block? Can you please post an example that is not working?

_Can skin be added to the well perforation model in future?_
Yes, this should be very easy. Please post a feature request.

**Reply by francoishamon** [reply]
@wdmhouston _Well perforation will be moved in parallel mode(bug). This causes issues with horizontal well_
Can you please post your xml? This happens because the "well mesh" (number of well segments) is not fine enough, there is a fix that I can show you

**Reply by wdmhouston** [reply]
Francois,
   
    See the flowing picture for vertical well and horizontal well. Is my understanding correct? 
![image](https://user-images.githubusercontent.com/19333848/122313825-8c7c1080-cedc-11eb-91f1-e455de25a09d.png)

for the example xml requested, I'll prepare an example next week.

**Reply by francoishamon** [reply]
Hi @wdmhouston 
Yes your understanding is correct, for the description of the well polyline and the location of the perforation.

The only point I want to mention is that the perforation is a point, not a segment, so there is no concept of a perforation length. Once the perforation point location has been deduced using the "distanceFromHead" keyword, GEOSX uses this location to introduce a connection between a well cell and a reservoir cell. The "strengh" of this connection is the well transmissibility.

If you checkout the latest version of `develop`, you can add `logLevel=1` to the `InternalWell` block to see what is the structure of the well mesh, and where the perforations are located:
https://github.com/GEOSX/GEOSX/blob/b61aaec5aaa97c0b19753464ed1c8b12efb976cc/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/staircase_co2_wells_3d.xml#L88-L102
